### DISCUSSION TITLE
For people who want to use DeoVR to play videos from Stash

### DISCUSSION THREAD

**Discussion by philpw99** [stashapp/stash#4777] [Show and tell]
I have been using my own fork of Stash designed for DeoVR for some months already, and I find it easy, fast and not rely on complicated setups. I think my release is good enough to share with you guys. It has minimum modification of original Stash, and it works well in Windows. I start the Stash in my Windows 10, and watch my VR videos with my Quest 2 running DeoVR.

So if you are interested, grab a copy here:
https://github.com/philpw99/stash4deovr/releases
and replace the stash-win.exe/stash-linux/stash-macos with the one in my fork ( That's all you need to do.).

This Stash should just work the same like the original. The only difference is : Now if you use DeoVR's browser to go to your Stash address, like "http://192.168.1.10:9999", click on one of the scene you want to watch, you will see something like this under the front cover:

![ext player](https://github.com/stashapp/stash/assets/22040708/f7c2b69b-295b-4f90-84d9-05edd5352350)

Now you can click on the "Open Ext." to play the VR video natively in DeoVR. DeoVR should know the projection format by the filename. If not, you can always tell DeoVR the project type just once, it will remember it next time.

DeoVR also remembers your progress for each VR video, therefore when you click on the "Open Ext." button again, it will jump to the place you watched last time.

When playing the video, you can click on the time bar to fast forward or go back. And if you click on the "Home" button, you will be back to the scene's web page. There you can go back and continue to browse your video collections.

Not only you can watch VR videos natively, you can watch any other 2D or 3D videos in your collection as long as DeoVR supports it. Note that DeoVR only supports Full SBS video, not the popular Half SBS ones.

You can also set your Stash's URL as the home page of DeoVR, it's in the DeoVR's setting. So next time you open DeoVR, it will open up the Stash page automatically.

The only problem so far, is the browser being too small. You can do the following to make it slightly better:
1. Map a key to zoom in the UI.
2. In the scene list, use "Wall" view instead of the regular "Grid" view.
3. Most of the time the front cover of a scene will cover the whole browser, you need to drag the page up to see the "Open Ext." button. You can use a custom CSS to make the front cover small. For example, ` .VideoPlayer { height: 200px } `

**Comment by philpw99** [comment]
I did try to do the PR. It seems my implement at that time had changed their structure too much. 
Also it seems they really don't want to host the file directly, for whatever the reason.
So in the end I had to created my own fork instead of waiting for approval.
After that I had suggested a few times Stash should have the direct file serving feature.
I haven't check Stash for months. Do they support direct file hosting now ? If they do I'm happy to create a PR.

**Comment by Bertrand-1010** [comment]
Hello, 

I tried to use your fork; But I obtain an error at startup : "error: Manager Initialization error: schema version 72 is incompatible with required schema version 56". Have you any idea?

Thanks,
TM

**Comment by philpw99** [comment]
The fork has been updated to 0.30.1 
And now there is an option switch in "Settings -> Interface -> Scene Player" to enable it or disable it.

**Comment by Bwolfiee206** [comment]
Typing in my ip address followed by :9999 i get 'webpage cannot be loaded because net::ERR_CONNECTION_REFUSED'. Not sure what i did wrong but i would love any tips or help if anyone's able. Thanks!

**Comment by philpw99** [comment]
I presume you are running stash-win.exe under Windows.
Most likely you need to run Windows Firewall by Run Command( Win + C ) and type "wf.msc" then hit Enter.
Then you add two inbound rules for this. One rule to allow program "stash-win.exe", one rule to allow TCP port 9999.
After the rules are set, you should be able to access Stash from other computer or device.

**Comment by mooierondjes-svg** [comment]
Hi this is great!

So I have sucessfully opened my lib on my quest 2.
I can start playing videos on my quest 2, however when I click "Open Ext." I get a "http response code failure" error.
So I can play the videos in non-vr. But opening the VR video is not possible

**Reply by philpw99** [reply]
Right now there is a new version compatible with Stash 0.30.1. This new version no longer has "Open Ext." words, instead just the button to open external file.
I just tested this version with Quest 3. It still works. The VR video will be opened in fullscreen mode.
The non-vr files were not important for this special Stash at all. This one was designed to play VR files with DeoVR natively.

**Reply by mindlesswind** [reply]
Thank you for your work. I have found similar solutions over the past few months.

Unfortunately, I tried using the latest version (v0.31.1-dev), but I still encountered the “http response code failure” error. I also noticed that the generated link appends /org/[video name] to the end of the URL.

When I manually remove that part, DeoVR is able to stream the video in flat-screen mode, but it still cannot play properly in full-screen VR mode.

I am using latest DeoVR version on quest 3.

**Reply by philpw99** [reply]
This is very strange because I have been using it this way for years.
The latest release doesn't care about the filename any more. It will just direct stream the scene. I think it might finally fix your problem?
The generated link is now /stream/[video name]
